prov/tcp/src/xnet_progress.c

```diff
@@ -1130,22 +1130,35 @@ static void xnet_progress_cqe(struct xnet_progress *progress,
 	struct fid *fid;
 	struct xnet_ep *ep;
 	struct xnet_conn_handle *conn;
+	struct xnet_pep *pep;
 
 	assert(xnet_io_uring);
 	sockctx = (struct ofi_sockctx *) cqe->user_data;
 	assert(sockctx);
 	assert(sockctx->uring_sqe_inuse);
-	sockctx->uring_sqe_inuse = false;
+	if (!(cqe ->flags & IORING_CQE_F_MORE))
+		sockctx->uring_sqe_inuse = false;
 	uring->sockapi->credits++;
 
 	fid = sockctx->context;
-	if (fid->fclass == FI_CLASS_EP) {
+	switch (fid->fclass) {
+	case FI_CLASS_EP:
 		ep = container_of(fid, struct xnet_ep, util_ep.ep_fid.fid);
 		xnet_uring_run_ep(ep, sockctx, cqe->res);
-	} else {
-		assert(fid->fclass == FI_CLASS_CONNREQ);
+		break;
+	case FI_CLASS_CONNREQ:
 		conn = container_of(fid, struct xnet_conn_handle, fid);
 		xnet_uring_run_conn(conn, cqe->res);
+		break;
+	default:
+		assert(fid->fclass == FI_CLASS_PEP);
+	case FI_CLASS_PEP:
+		pep = container_of(fid, struct xnet_pep, util_pep.pep_fid.fid);
+		if (sockctx == &pep->pollin_sockctx) {
+		    if (cqe->res >= 0)
+				xnet_accept_sock(pep);
+		} else
```

Comment:
nit: need braces around else

Reply:
> Do we reserve space in the iouring for the management requests (cancel, pollin, pollout)?
No we don't reserve space for management requests. That's definitely a problem with the current design if the ring runs out of credits. I'm not sure exactly how to gracefully address this problem.

Reply:
Can we reduce credits by 4 (or whatever) and track those as mgmt_credits?  Though I'm not sure we even need to track that, versus just tracking that we only have 1 mgmt request per type outstanding at a time.

Reply:
The problem is that the ring is shared between all endpoints from the same domain. If we run out of `mgmt_credits` and an endpoint needs to submit a new `mgmt` SQE (pollin, pollout or cancel), it will need to wait for a new credit to become available. Before this PR, the code was relying on epoll to retry an endpoint when the ring runs out of credits. But since we are moving epoll operations to the ring, this solution no longer works. 

I see several options here:
- Provision enough entries in the ring in order to make sure it never overflows.
- Leave the responsibility to the application to guarantee that no overflow would occur on the ring. If an overflow occurs, endpoints will be automatically shut down and disabled. This option is a similar to the Completion Queues with Infiniband where an overflow would transition all associated endpoints to an error state.
- Each domain has a pool of rings, which can grow on-demand.

Reply:
I don't see how we can make the app responsible for this.  With CQs, the app at least knows that there's a completion per request, so it has a way to track it.  But pollin/pollout are internal and/or related to the remote side.  It seems like the current solution would be to always allocate a giant ring and hope it's big enough.  I know there have been recent discussions on the kernel mailing list about having the ability to resize the rings, specifically because network apps can't guess the needed size up front.  Hopefully we can figure out some way to keep things running smoothly in cases where the ring does get full (and assuming that progress is being made on completing some number of queued requests).

Reply:
Avoiding overrunning io_uring is somewhat equivalent to avoiding overrunning the verbs CQ.  So, maybe this isn't that big of a deal, at least I think it's the same verbs would have.

Reply:
Hi @shefty, I’ve been thinking a bit more about that io_uring overrunning issue the last couple of days. Here are my conclusions:

When the provider issues an operation but the ring is running out of credits, we could submit all pending SQEs, reap the ready CQEs and save them into a bufpool until one credit gets released. This solution is however not perfect because there is absolutely no guarantee that a single CQE will be released if the ring is full. For example, a posted receive buffer may be released only if a packet is received, the remote process is dead or the corresponding SQE is canceled.

The credit system we have implemented in the TCP provider guarantees that the SQ never overruns. With io_uring, the SQE is immediately released after the kernel has consumed the request. Moreover, with relatively recent Linux kernels, io_uring guarantees that a single CQ almost never drops events in case of overflow (unless the kernel runs out of memory). If the overflow condition is reached, submitting new SQEs would fail with -EBUSY.

I’ve reviewed a couple of applications supporting io_uring and some of them (e.g., tokio-rs https://github.com/tokio-rs/tokio-uring) seem to completely rely on the capacity of io_uring to overflow. It also seems to be the solution recommended by Jens (see https://github.com/axboe/liburing/issues/396)
If get_sqe returns NULL, then the SQ is overflowed and ready SQEs need to be submitted. SQEs that are consumed by the kernel are immediately released and become available to the application. If submitting SQEs fails with -EBUSY, then the CQ is overflowed and the application needs to poll the CQ and reap CQEs. Overflowing the queues may significantly impact the overflow performance, but at least the application should continue running if the number of ring entries is underestimated by the upper layer.

I’ve implemented the above solution in libfabric (i.e., removed the ring credits from the TCP provider and let the ring overflow), but the code doesn’t work well. It seems that the version of the Linux kernel I’m using for development doesn’t retry overflowed CQEs and, as a result, CQEs get processed out of order. This problem is described here: https://lore.kernel.org/io-uring/20200707132420.2007-1-xiaoguang.wang@linux.alibaba.com/

For the moment, it’s probably safer to keep our own ring credits in the TCP provider and reap SQEs if all the credits are exhausted. Reaped SQEs would be copied into a bufpool entry and added into a list that would be processed when polling the ring, similarly to the verbs provider when the CQ is flushed. What do you think?

Reply:
It should work okay in practice to track credits ourselves.  It's easy to fail transmit requests with EAGAIN if there aren't credits available.  Receives are another matter, as that could lead to deadlock on large systems.  As long as the number of credits is > 2 x # peers, we should be okay.  We just need a large enough queue to support the maximum number of peers that we expect to support.  I would expect that to be in the low thousands.